src/psyclone/psyir/transformations/hoist_loop_bound_expr_trans.py

```diff
@@ -0,0 +1,148 @@
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: S. Siso, STFC Daresbury Lab
+
+'''This module contains the HoistLoopBoundExprTrans transformation. This
+transformation moves complex bounds expressions out of the loop construct and
+places them in integer scalar assignments before the loop.
+
+'''
+
+from psyclone.psyir.nodes import Routine, Literal, Reference, \
+    StructureReference, Assignment
+from psyclone.psyir.symbols import DataSymbol, INTEGER_TYPE
+from psyclone.psyir.transformations.loop_trans import LoopTrans, \
+    TransformationError
+
+
+class HoistLoopBoundExprTrans(LoopTrans):
+    '''This transformation moves complex bounds expressions out of the loop
+    construct and places them in integer scalar assignments before the loop.
+
+    >>> from psyclone.psyir.backend.fortran import FortranWriter
+    >>> from psyclone.psyir.frontend.fortran import FortranReader
+    >>> from psyclone.psyir.nodes import Loop
+    >>> from psyclone.psyir.transformations import HoistTrans
+    >>> code = ("program test\\n"
+    ...         "  use mymod, only: mytype\\n"
+    ...         "  integer :: i,j,n\\n"
+    ...         "  real :: a(n)\\n"
+    ...         "  do i=mytype%start, UBOUND(a,1)\\n"
+    ...         "    a(i) = 1.0\\n"
+    ...         "  end do\\n"
+    ...         "end program\\n")
+    >>> psyir = FortranReader().psyir_from_source(code)
+    >>> hoist = HoistLoopBoundExprTrans()
+    >>> hoist.apply(psyir.walk(Loop)[0])
+    >>> print(FortranWriter()(psyir))
+    program test
+      integer :: i
+      integer :: j
+      integer :: n
+      real, dimension(n,n) :: a
+      real :: value
+    <BLANKLINE>
+      value = 1.0
+      do i = 1, n, 1
+        do j = 1, n, 1
+          a(i,j) = value
+        enddo
+      enddo
+    <BLANKLINE>
+    end program test
+    <BLANKLINE>
+
+    '''
+    def apply(self, node, options=None):
+        '''Move complex bounds expressions out of the given loop construct and
+        place them in integer scalar assignments before the loop.
+
+        :param node: target PSyIR loop.
+        :type node: :py:class:`psyclone.psyir.nodes.Loop`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        '''
+        self.validate(node, options)
+
+        parent = node.parent
+        position = node.position
+
+        for bound in node.children[0:3]:
```

Comment:
Although I like the simplicity of this approach I do wonder whether we should use the semantic navigation (i.e. `node.start_expr`) and then we could give the new symbols more meaningful names too. Is it possible to do that without too much duplication? Perhaps:

    for (name, bound) in [("start", node.start_expr), ("stop", node.stop_expr), ...]:]

Reply:
Updated as suggested